subsys/caf/modules/sensor_sampler.c

```diff
@@ -0,0 +1,206 @@
+		data_ptr[i] = sensor_value_to_double(&data[i]);
+	}
+
+	EVENT_SUBMIT(event);
+}
+
+static int sample_sensor(const struct device *dev, const struct sensor_config *sc)
+{
+	size_t data_cnt = 0;
+
+	for (size_t i = 0; i < sc->chan_cnt; i++) {
+		data_cnt += sc->chans[i].data_cnt;
+	}
+
+	int err = 0;
+	size_t data_idx = 0;
+	struct sensor_value data[data_cnt];
+
+	for (size_t i = 0; !err && (i < sc->chan_cnt); i++) {
+		const struct sampled_channel *sampled_chan = &sc->chans[i];
+
+		err = sensor_sample_fetch_chan(dev, sampled_chan->chan);
+		if (!err) {
+			err = sensor_channel_get(dev, sampled_chan->chan, &data[data_idx]);
+			data_idx += sampled_chan->data_cnt;
+		}
+	}
+
+	if (err) {
+		LOG_ERR("Sensor sampling error (err %d)", err);
+	} else {
+		send_sensor_event(sc->event_descr, data, ARRAY_SIZE(data));
+	}
+
+	return err;
+}
+
+static int sample_sensors(int *wait_ms)
+{
+	int err = 0;
+	int64_t delta = k_uptime_delta(&uptime);
+
+	*wait_ms = INT_MAX;
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		sensor_data[i].remaining_wait -= delta;
+
+		if (sensor_data[i].remaining_wait <= 0) {
+			err = sample_sensor(sensor_data[i].dev, &sensor_configs[i]);
+
+			while (sensor_data[i].remaining_wait <= 0) {
+				sensor_data[i].remaining_wait += sensor_data[i].sampling_period;
+			}
+
+			/* Run sampling again if any of the sensors was sampled. */
+			*wait_ms = 0;
+		}
+
+		if (*wait_ms > sensor_data[i].remaining_wait) {
+			*wait_ms = sensor_data[i].remaining_wait;
+		}
+	}
+
+	return err;
+}
+
+static int sensor_init(void)
+{
+	int err = 0;
+	struct sensor_data *sd;
+	const struct sensor_config *sc;
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		sd = &sensor_data[i];
+		sc = &sensor_configs[i];
+
+		sd->dev = device_get_binding(sc->dev_name);
+		if (!sd->dev) {
+			err = -ENXIO;
+			break;
+		}
+		sd->sampling_period = sc->sampling_period_ms;
+		sd->remaining_wait = sc->sampling_period_ms;
+	}
+
+	if (!err) {
+		state = STATE_ACTIVE;
+		module_set_state(MODULE_STATE_READY);
+		uptime = k_uptime_get();
+	} else {
+		report_error();
+	}
+
+	return err;
+}
+
+static void sample_thread_fn(void)
+{
+	int err = 0;
+	int sleep_ms;
+
+	err = sensor_init();
+
+	while (!err) {
+		err = sample_sensors(&sleep_ms);
+
+		if (sleep_ms == INT_MAX) {
+			k_sleep(K_FOREVER);
+		} else {
+			k_sleep(K_MSEC(sleep_ms));
+		}
+	}
+
+	report_error();
+}
+
+static int init(void)
+{
+	k_thread_create(&sample_thread, sample_thread_stack, SAMPLE_THREAD_STACK_SIZE,
+			(k_thread_entry_t)sample_thread_fn, NULL, NULL, NULL,
+			SAMPLE_THREAD_PRIORITY, 0, K_NO_WAIT);
+	k_thread_name_set(&sample_thread, "caf_sensor_sampler");
+
+	return 0;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_module_state_event(eh)) {
+		struct module_state_event *event = cast_module_state_event(eh);
+
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			__ASSERT_NO_MSG(state == STATE_DISABLED);
+
+			init();
```

Comment:
if error from init is not handled why to bother returning the value?

Reply:
Leftover from previous implementation that initialized sensor devices here. Fixed